board/safety/safety_tesla.h

```diff
@@ -0,0 +1,305 @@
+    uint32_t ts_elapsed = get_ts_elapsed(ts, tesla_ts_angle_last);
+
+    // *** angle real time check
+    // add 1 to not false trigger the violation and multiply by 25 since the check is done every 250 ms and steer angle is updated at     100Hz
+    int rt_delta_angle_up = ((int)((interpolate(TESLA_LOOKUP_ANGLE_RATE_UP, tesla_speed) * 25. + 1.)));
+    int rt_delta_angle_down = ((int)((interpolate(TESLA_LOOKUP_ANGLE_RATE_DOWN, tesla_speed) * 25. + 1.)));
+    int highest_rt_angle = tesla_rt_angle_last + (tesla_rt_angle_last > 0 ? rt_delta_angle_up : rt_delta_angle_down);
+    int lowest_rt_angle = tesla_rt_angle_last - (tesla_rt_angle_last > 0 ? rt_delta_angle_down : rt_delta_angle_up);
+
+    if ((ts_elapsed > TESLA_RT_INTERVAL) || (controls_allowed && !tesla_controls_allowed_last))
+    {
+      tesla_rt_angle_last = angle_meas_now;
+      tesla_ts_angle_last = ts;
+    }
+
+    // update array of samples
+    update_sample(&tesla_angle_meas, angle_meas_now);
+
+    // check for violation;
+    if (max_limit_check(angle_meas_now, highest_rt_angle, lowest_rt_angle))
+    {
+      // We should not be able to STEER under these conditions
+      // Other sending is fine (to allow human override)
+      steer_allowed = 0;
+      puts("WARN: RT Angle - No steer allowed! \n");
+    }
+    else
+    {
+      steer_allowed = 1;
+    }
+
+    tesla_controls_allowed_last = controls_allowed;
+  }
+}
+
+// all commands: gas/regen, friction brake and steering
+// if controls_allowed and no pedals pressed
+//     allow all commands up to limit
+// else
+//     block all commands that produce actuation
+
+static int tesla_tx_hook(CAN_FIFOMailBox_TypeDef *to_send)
+{
+
+  uint32_t addr;
+  int angle_raw;
+  int desired_angle;
+
+  addr = to_send->RIR >> 21;
+
+  // do not transmit CAN message if steering angle too high
+  // DAS_steeringControl::DAS_steeringAngleRequest
+  if (addr == 0x488)
+  {
+    angle_raw = ((to_send->RDLR & 0x7F) << 8) + ((to_send->RDLR & 0xFF00) >> 8);
+    desired_angle = angle_raw * 0.1 - 1638.35;
+    int16_t violation = 0;
+    int st_enabled = (to_send->RDLR & 0x400000) >> 22;
+
+    if (st_enabled == 0) {
+      //steering is not enabled, do not check angles and do send
+      tesla_desired_angle_last = desired_angle;
+      return true;
+    }
+
+    if (controls_allowed)
+    {
+      if (steer_allowed)
+      {
+
+        // add 1 to not false trigger the violation
+        int delta_angle_up = (int)(interpolate(TESLA_LOOKUP_ANGLE_RATE_UP, tesla_speed) * 25. + 1.);
+        int delta_angle_down = (int)(interpolate(TESLA_LOOKUP_ANGLE_RATE_DOWN, tesla_speed) * 25. + 1.);
+        int highest_desired_angle = tesla_desired_angle_last + (tesla_desired_angle_last > 0 ? delta_angle_up : delta_angle_down);
+        int lowest_desired_angle = tesla_desired_angle_last - (tesla_desired_angle_last > 0 ? delta_angle_down : delta_angle_up);
+        int TESLA_MAX_ANGLE = (int)(interpolate(TESLA_LOOKUP_MAX_ANGLE, tesla_speed) + 1.);
```

Comment:
we also need a simple check that ensures the rate limits are observed at each step (I only see the real time check). See safety_gm.h, for example.